internal/app/export/service.go

```diff
@@ -4,62 +4,294 @@
+		Logger:    l,
+		Cfg:       c,
+		Consumers: routes,
+		Cache:     cache,
+		connected: make(chan bool, 1),
 	}
-
+	client, err := e.mqttConnect(c, l)
+	if err != nil {
+		log.Fatalf(err.Error())
+	}
+	e.MQTT = client
 	return &e
 }
 
-// Start method starts consuming messages received from NATS
-// and makes routes according to the configuration file.
-// Routes export messages to mqtt.
-func (e *exporter) Start(queue string, nc *nats.Conn) {
+// Start method loads route configuration
+func (e *exporter) Start(queue string) error {
 	var route routes.Route
 	for _, r := range e.Cfg.Routes {
+		natsTopic := fmt.Sprintf("%s.%s", NatsSub, r.NatsTopic)
 		switch r.Type {
 		case "mfx":
-			route = mfx.NewRoute(r.NatsTopic, r.MqttTopic, r.SubTopic, e.Mqtt, e.Logger)
+			route = mfx.NewRoute(natsTopic, r.MqttTopic, r.SubTopic)
 		default:
-			route = routes.NewRoute(r.NatsTopic, r.MqttTopic, r.SubTopic, e.Mqtt, e.Logger)
+			route = routes.NewRoute(natsTopic, r.MqttTopic, r.SubTopic)
+		}
+		if !e.validateSubject(route.NatsTopic()) {
+			continue
+		}
+		e.Consumers[route.NatsTopic()] = route
+		if e.Cache != nil {
+			g, err := e.Cache.GroupCreate(r.NatsTopic, exportGroup)
+			if err != nil {
+				e.Logger.Error(fmt.Sprintf("Failed to create stream group %s", err.Error()))
+			}
+			e.Logger.Info(fmt.Sprintf("Stream group %s created %s", r.NatsTopic, g))
+		}
+
+	}
+	if len(e.Consumers) == 0 {
+		return errNoRoutesConfigured
+	}
+	if e.Cache != nil {
+		go e.startRepublish()
+	}
+	return nil
+}
+
+func (e *exporter) Consume(msg *nats.Msg) {
+	if _, ok := e.Consumers[msg.Subject]; !ok {
+		e.Logger.Info(fmt.Sprintf("no configuration for nats topic %s", msg.Subject))
+		return
+	}
+
+	route := e.Consumers[msg.Subject]
+	payload, err := route.Consume(msg)
+	if err != nil {
+		e.Logger.Error(fmt.Sprintf("Failed to consume msg %s", err.Error()))
+	}
+
+	if err = e.publish(route.MqttTopic(), payload); err != nil {
+		if e.Cache == nil {
+			return
+		}
+		// If error occurred we will store data to try to republish
+		_, err := e.Cache.Add(msg.Subject, route.MqttTopic(), payload)
+		if err != nil {
+			e.Logger.Error(fmt.Sprintf("Failed to add to redis stream `%s`", msg.Subject))
 		}
-		e.Routes = append(e.Routes, route)
-		e.Subscribe(route, queue, nc)
 	}
 }
 
-func (e *exporter) Subscribe(r routes.Route, queue string, nc *nats.Conn) {
-	_, err := nc.QueueSubscribe(r.NatsTopic(), fmt.Sprintf("%s-%s", queue, r.NatsTopic()), r.Consume)
+func (e *exporter) startRepublish() {
+	// Initial connection established on start up
+	<-e.connected
+	e.Logger.Info("Starting republish, waiting for stream data")
+	for _, route := range e.Cfg.Routes {
+		streams := []string{route.NatsTopic, ">"}
+		go func() {
+			for {
+				msgs, err := e.readMessages(streams)
+				if err != nil {
+					continue
+				}
+				e.Logger.Info(fmt.Sprintf("Waiting for connection to %s", e.Cfg.MQTT.Host))
+				for {
+					// Wait for connection
+					if e.IsConnected() || <-e.connected {
```

Comment:
Can you use only e.IsConnected() (I assume that's the purpose of that method)?

Reply:
nop, because it has to wait here untill gets connected